### ISSUE TITLE
Adjust patience and other NER training duration parameters

### ISSUE THREAD

**Issue by FrancescoCasalegno** [BlueBrain/Search#356]
We are currently training our NER models using the default parameters of the `config.cfg` for the duration and early stopping:
https://github.com/BlueBrain/Search/blob/04aa04bb68ecca7222046bdeb05bafbee9e01665/data_and_models/pipelines/ner/config.cfg#L88-L90

But, as discussed in https://github.com/BlueBrain/Search/issues/343#issuecomment-831314232, we can probably set `patience` and `max_steps` to much smaller values without any negative impact on the accuracy of our models.

Also, we should look into the learning rate schedule parameters that should probably also be adapted.

**Comment by pafonta** [comment]
Hello @FrancescoCasalegno,

> Also, we should look into the learning rate schedule parameters that should probably also be adapted.

Here is the section in `config.cfg` (from the PR introducing the `transformer` component):
https://github.com/BlueBrain/Search/blob/f855db2871a82f4b1f4723b32fc2845658571887/data_and_models/pipelines/ner/config.cfg#L136-L140

The documentation for `warmup_linear.v1` is there:
https://thinc.ai/docs/api-schedules#warmup_linear

If one wants to use a `fixed` learning rate, one could do this by `deleting` the section `[training.optimizer.learn_rate]` and `adding` the parameter `learn_rate` as part of the section `[training.optimizer]`.

The documentation for `Adam.v1` is there:
https://thinc.ai/docs/api-optimizers#adam

Note:
Full documentation links for `config.cfg`: https://github.com/BlueBrain/Search/pull/328#issuecomment-816679707.

**Comment by pafonta** [comment]
I am having a look into this issue.

There are 2 things to consider, I think:

1. We use in #335 the `same` training parameters as `spaCy` for their SOTA model, `en_core_web_trf`. So we might want to be careful about changing them.

2. If we use the scores from the training to select the new parameters, we might have `erroneous` conclusions. Indeed:
  a. We currently use the scores from `our` evaluation and on the `test` set to select the best NER approach.
  b. We know that `spaCy` evaluation and our evaluation `doesn't` compare (see #321).

**Comment by pafonta** [comment]
## Experiment

Goal:
- If we use a `constant` learning rate and train for a `fixed` and large number of steps / epochs, could we find from the collected data **which configuration would reduce training time while keeping the current accuracy**?

<details>
  <summary>setup</summary>

Base:
- HEAD of #351 (906d95deabc2ee2ff30d920b9e9bbceac9c16bc7).

Modifications:
- Use a `patience` of 0. The training is then done for `max_steps`, i.e. `20,000`.
- Use a `constant` learning rate of 5e-5, instead of a `warmup_linear` schedule (which has an initial learning rate of 5e-5).

</details>

## Observations

1. As expected in point 2 above (https://github.com/BlueBrain/Search/issues/356#issuecomment-838479923), we `cannot` make a `correct` decision on a number of maximum steps / epochs until #321 is sorted out. Indeed, the performances on the dev set and the test set could be in `contradiction`. For example, in the table below, for entity chemical, the `dev` F1 score `improved` by ~30 %  while the `test` F1 score `decreased` by ~ 10 %. This could be over-fitting but, at the same time, this contradiction isn't there for all the entities. So, there should be another or an additional cause than over-fitting.

2. It doesn't seem there is a correlation between the best dev F1 and the number of steps or epochs. Therefore, we `cannot` say from the experiment data what would be the `best trade-off` between `training time` (i.e. number of steps / epochs) and the `performances` (i.e. F1 score).

## Results

<details>
  <summary>summary</summary>

entity | first epoch when best dev F1 is reached | best dev F1 | relative delta compared to #351 for best dev F1 | test F1 | relative delta compared to #351 for test F1
-- | -- | -- | -- | -- | --
cell_compartment | 1400 | 0.8525 | 3.28 | 0.8000 | 2.71
cell_type | 400 | 0.6400 | 5.14 | 0.7059 | 3.02
chemical | 1667 | 0.6842 | 28.95 | 0.4400 | -12.63
drug | 29 | 0.8333 | 4.16 | 0.7044 | 6.49
organism | 1400 | 0.8966 | 0.48 | 0.5972 | -8.66
protein | 5200 | 0.8182 | 22.73 | 0.4783 | -15.12

➡️ Entities `disease`, `organ`, and `pathway` were not considered due to time constraints.

</details>

**Comment by pafonta** [comment]
TODO

However, thanks to the collected data, we could have an idea of the minimal number of steps / epochs needed to reach the current accuracy.

**Comment by pafonta** [comment]
## Validation

⚠️ WORK IN PROGRESS

Using the `proposed` solution above (https://github.com/BlueBrain/Search/issues/356#issuecomment-842452733):
- The training `time` was FIXME instead of 22.87 hours. That's a relative `gain` of FIXME %.
- The resulting `accuracy` is in `average`:
  - FIXME on the `dev` set instead of 0.7206. That's a relative `FIXME` of FIXME %.
  - FIXME on the `test` set instead of 0.6411. That's a relative `FIXME` of FIXME %.

➡️ In conclusion:
1. We can `FIXME` the proposed solution.
2. We can even lower the `patience` to FIXME steps and the `max_steps` to FIXME epochs.

<details>
  <summary>details for the _dev_ set</summary>

entity | epoch of F1 | step of F1 | F1 | current F1 | relative delta compared to   current | best F1 | relative delta compared to best
-- | -- | -- | -- | -- | -- | -- | --
cell_compartment | 49 | 400 | 0.8197 | 0.8254 | -0.69 | 0.8525 | -3.85
cell_type | 266 | 800 | 0.6400 | 0.6087 | 5.14 | 0.6400 | 0.00
chemical | 122 | 1200 | 0.6512 | 0.5306 | 22.73 | 0.6842 | -4.82
drug |  |  |  | 0.8000 |  | 0.8333 | 
organism |  |  |  | 0.8923 |  | 0.8966 | 
protein | 333 | 1000 | 0.7273 | 0.6667 | 9.09 | 0.8182 | -11.11

</details>

<details>
  <summary>details for the _test_ set</summary>

entity | F1 | current F1 | relative delta compared to   current
-- | -- | -- | --
cell_compartment |  | 0.7789 | 
cell_type |  | 0.6852 | 
chemical |  | 0.5036 | 
drug |  | 0.6615 | 
organism |  | 0.6538 | 
protein |  | 0.5635 | 

</details>

**Comment by pafonta** [comment]
## Supplementary material

I had made some `plots` which I haven't included yet to make the results more digestible.

However, I think they are of interest and it might be worth saving them here for further reference.

They could also be useful for #321.

Here are the plots:

### best dev F1 vs epochs

<details>
  <summary>plot</summary>

![best dev F1 vs epochs](https://user-images.githubusercontent.com/4543376/119116411-80f61080-ba28-11eb-85a7-88d94e5bf711.png)

</details>

### dev F1 & losses vs epochs

➡️ The dotted vertical black `line` is the epoch for which the best dev F1 is reached. It's exact `value` is in the plot `title`.

<details>
  <summary>cell_compartment</summary>

![cell_compartment](https://user-images.githubusercontent.com/4543376/119116785-e1854d80-ba28-11eb-913c-05d773a94c87.jpg)

</details>

<details>
  <summary>cell_type</summary>

![cell_type](https://user-images.githubusercontent.com/4543376/119117085-2b6e3380-ba29-11eb-8879-c5a5b4680ab0.jpg)

</details>

<details>
  <summary>chemical</summary>

![chemical](https://user-images.githubusercontent.com/4543376/119117166-42148a80-ba29-11eb-9ac5-f93d253469ed.jpg)

</details>

<details>
  <summary>drug</summary>

![drug](https://user-images.githubusercontent.com/4543376/119117207-4b9df280-ba29-11eb-851d-11a5dfdfd0fa.jpg)

</details>

<details>
  <summary>organism</summary>

![organism](https://user-images.githubusercontent.com/4543376/119117250-59ec0e80-ba29-11eb-8403-0344298bd2b6.jpg)

</details>

<details>
  <summary>protein</summary>

![protein](https://user-images.githubusercontent.com/4543376/119117283-62dce000-ba29-11eb-8f55-a58eedf68f32.jpg)

</details>
